### PR TITLE
feat: add type hints and mypy

### PR DESCRIPTION
**Pull request by peterdeme**
## Features

- Most importantly added type hints everywhere.
- Enforcing typing during the build with `mypy`
- removed docstrings from derived classes. docstrings are automatically inherited from the abstract class ‼️ no need to declare them twice.
- adding [py.typed file](https://www.python.org/dev/peps/pep-0561/) to the package
- removed unnecessary dependencies
- made testing logic much simpler. it was a bit weird, what was that long thing in setup.py? :)
- fix repository url in setup.py
- don't include test files in the published package ❗ 
- applying more `flake8` rules. Plus fixed two flake8 issues (`__all__` and `assert False`)
- added `flake8-bugbear` package, one of my favourite bug finder 🙂 
- updated release notes: let's upload wheels as well, not just sdists. wheels are generally better.
- added reviewdog

## Future improvements to come
- switch from `setup.py` + `MANIFEST.in` fully over to `pyproject.toml`. we can unify these 3 files into one. I worked a lot with this in the previous company.
- introduce package publishing with GitHub Actions

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,3 +23,16 @@ exclude = '''
...
 '''
+
+[tool.mypy]
+disallow_untyped_calls = true
+check_untyped_defs = true
+warn_unused_configs = true
+
+[[tool.mypy.overrides]]
+module = "aiofile.*"
+ignore_missing_imports = true
```
Comment (peterdeme) [inline pyproject.toml:34]: https://github.com/mosquito/aiofile/pull/62

File: MANIFEST.in [hunk 2]
```diff
@@ -1 +1,2 @@
-include README.md
\ No newline at end of file
+include README.md
+include stream_chat/py.typed
```
Comment (ferhatelmas) [inline MANIFEST.in:2]: eol

File: README.md [hunk 3]
```diff
@@ -125,7 +127,7 @@ In order to release new version you need to be a maintainer on Pypi.
 - Update the version on setup.py
 - Commit and push to Github
 - Create a new tag for the version (eg. `v2.9.0`)
-- Create a new dist with python `python setup.py sdist`
-- Upload the new distributable with twine `twine upload dist/stream-chat-VERSION-NAME.tar.gz`
+- Create a new dist with python `python setup.py sdist bdist_wheel`
+- Upload the new distributable with twine `twine upload dist/*`
```
Comment (ferhatelmas) [inline README.md:131]: 👍🏻

File: setup.py [hunk 4]
```diff
@@ -1,61 +1,41 @@
-import sys
-
 from setuptools import find_packages, setup
-from setuptools.command.test import test as TestCommand
 
 install_requires = [
-    "pycryptodomex>=3.8.1,<4",
```
Comment (ferhatelmas) [inline setup.py:7]: they were leftover from pyjwt 1.x

File: stream_chat/__pkg__.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
 __author__ = "Tommaso Barbugli"
-__copyright__ = "Copyright 2019-2021, Stream.io, Inc"
+__copyright__ = "Copyright 2019-2022, Stream.io, Inc"
```
[the remark this answers is not in the document]
Comment (peterdeme) [reply stream_chat/__pkg__.py:2]: 🎅

**Review by ferhatelmas — DISMISSED** [review]
LGreatTM

File: pyproject.toml (hunk 1 above)
Comment (peterdeme) [reply pyproject.toml]: this is now resolved

File: stream_chat/channel.py [hunk 6]
```diff
@@ -1,74 +1,41 @@
...
 
-    def send_event(self, event, user_id):
-        """
-        Send an event on this channel
-
-        :param event: event data, ie {type: 'message.read'}
-        :param user_id: the ID of the user sending the event
-        :return: the Server Response
-        """
+    def send_event(self,     event: Dict, user_id: str) -> StreamResponse:
```
[the remark this answers is not in the document]
Comment (peterdeme) [reply stream_chat/channel.py]: This was a test ☝️

File: stream_chat/channel.py [hunk 7]
```diff
@@ -1,74 +1,41 @@
...
         payload = {"event": add_user_id(event, user_id)}
-        return self.client.post(f"{self.url}/event", data=payload)
-
-    def send_reaction(self, message_id, reaction, user_id):
-        """
-        Send a reaction about a message
+        t = self.client.post(f"{self.url}/event", data=payload)
+        return 1
```
[the remark this answers is not in the document]
Comment (peterdeme) [reply stream_chat/channel.py]: This was a test ☝️

**Comment by ferhatelmas** [comment]
@peterdeme all good here, feel free to merge when your tests are done

**Review by ferhatelmas — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -33,6 +37,5 @@ jobs:
         env:
           STREAM_KEY: ${{ secrets.STREAM_KEY }}
           STREAM_SECRET: ${{ secrets.STREAM_SECRET }}
-        run: |
-          python setup.py install
-          make test
+          PYTHONPATH: ${{ github.workspace }}
```
Comment (ferhatelmas) [inline .github/workflows/ci.yml:40]: adding pip to path might not be needed anymore, good to update setup python

**Comment by peterdeme** [comment]
@ferhatelmas I can't, it requires @mahboubii or @gumuz 's approval.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (peterdeme) [reply .github/workflows/ci.yml:36]: let me see

**Review by gumuz — APPROVED** [review]
it's huge, so I didn't check all type-stuff, ut LGTM

File: .github/workflows/ci.yml (hunk 8 above)
Comment (peterdeme) [reply .github/workflows/ci.yml:36]: indeed. it works just fine without it. this `PYTHONPATH` is needed for pytest though. 🤷

**Pull request merged**